routes/studentRoutes.js

```diff
@@ -256,29 +260,60 @@ async function markExerciseAsCompleted(user, courseId, sectionId, exerciseId) {
           $push: {
             [`completedCourses.${completedCourseIndex}.completedSections`]: {
               sectionId,
-              completedExercises: [{ exerciseId }],
-              isComplete: false
+              completedExercises: [{ exerciseId, isComplete: isComplete, pointsGiven: points }]
             }
           }
         }
       );
     } else {
-      const isExerciseAlreadyCompleted = user.completedCourses[completedCourseIndex].completedSections.some(completedSection =>
-        completedSection.sectionId.equals(sectionId) &&
-        completedSection.completedExercises.some(completedExercise => completedExercise.exerciseId.equals(exerciseId))
-      );
-
-      // Returns if the exercise is already completed
-      if (isExerciseAlreadyCompleted) {
-        throw errorCodes['E0801'];
-      }
+      let exerciseFound;
```

Comment:
These functions should have a complete rework at some point - they are not the most readable 😄 

But let's add it to our technical debt 👍

Reply:
I agree